program/tests/namespace_royalty_enforcement.rs

```diff
@@ -0,0 +1,1123 @@
+#![cfg(feature = "test-bpf")]
+
+pub mod utils;
+
+use std::collections::HashSet;
+
+use mpl_token_auth_rules::{
+    error::RuleSetError,
+    instruction::{builders::ValidateBuilder, InstructionBuilder, ValidateArgs},
+    payload::{Payload, PayloadType},
+    state::{CompareOp, Rule, RuleSetV1},
+};
+use solana_program::{instruction::InstructionError, pubkey, pubkey::Pubkey};
+use solana_program_test::{tokio, ProgramTestContext};
+use solana_sdk::{
+    instruction::AccountMeta,
+    signature::Signer,
+    signer::keypair::Keypair,
+    system_instruction,
+    transaction::{Transaction, TransactionError},
+};
+use utils::{
+    create_associated_token_account, create_mint, program_test, DelegateScenario,
+    MetadataDelegateRole, Operation, PayloadKey, TokenDelegateRole, TransferScenario,
+};
+
+const ADDITIONAL_COMPUTE: u32 = 400_000;
+const RULE_SET_NAME: &str = "Metaplex Royalty RuleSet Dev";
+
+// --------------------------------
+// Define Program Allow List
+// --------------------------------
+const ROOSTER_PROGRAM_ID: Pubkey = pubkey!("Roostrnex2Z9Y2XZC49sFAdZARP8E4iFpEnZC5QJWdz");
+const TOKEN_METADATA_PROGRAM_ID: Pubkey = pubkey!("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s");
+const TOKEN_AUTH_RULES_ID: Pubkey = pubkey!("auth9SigNpDKz4sJJ1DfCTuZrZNSAgh9sFD3rboVmgg");
+
+const TRANSFER_PROGRAM_BASE_ALLOW_LIST: [Pubkey; 3] = [
+    TOKEN_METADATA_PROGRAM_ID,
+    ROOSTER_PROGRAM_ID,
+    TOKEN_AUTH_RULES_ID,
+];
+const DELEGATE_PROGRAM_BASE_ALLOW_LIST: [Pubkey; 3] = [
+    TOKEN_METADATA_PROGRAM_ID,
+    ROOSTER_PROGRAM_ID,
+    TOKEN_AUTH_RULES_ID,
+];
+const ADVANCED_DELEGATE_PROGRAM_BASE_ALLOW_LIST: [Pubkey; 3] = [
+    TOKEN_METADATA_PROGRAM_ID,
+    ROOSTER_PROGRAM_ID,
+    TOKEN_AUTH_RULES_ID,
+];
+
+struct ComposedRules {
+    transfer_rule: Rule,
+    wallet_to_wallet_rule: Rule,
+    delegate_rule: Rule,
+    advanced_delegate_rule: Rule,
+    namespace_rule: Rule,
+}
+
+// Get the four Composed Rules used in this RuleSet.
+fn get_composed_rules() -> ComposedRules {
+    // --------------------------------
+    // Create Primitive Rules
+    // --------------------------------
+    let nft_amount = Rule::Amount {
+        field: PayloadKey::Amount.to_string(),
+        amount: 1,
+        operator: CompareOp::Eq,
+    };
+
+    // Generate some random programs to add to the base lists.
+    let random_programs = (0..18).map(|_| Keypair::new().pubkey()).collect::<Vec<_>>();
+
+    let source_program_allow_list = Rule::ProgramOwnedList {
```

Comment:
Noting this for future reference, I believe at these quantity of pubkey list we were seeing 50k compute increase when doing hashset.